test/runnable/test8006.d

```diff
@@ -0,0 +1,402 @@
+
+    @property string x()
+    {
+        return mX;
+    }
+
+    @property void x(string v)
+    {
+        mX = v;
+    }
+
+    alias x this;
+}
+
+// same as TString, but setter @property function returns a value
+struct TStringRet
+{
+    string mX;
+
+    @property string x()
+    {
+        return mX;
+    }
+
+    @property string x(string v)
+    {
+        mX = v;
+        return mX;
+    }
+
+    alias x this;
+}
+
+// same as TString, but for static @property functions
+struct TStringStatic
+{
+    static string mX;
+
+    static @property string x()
+    {
+        return mX;
+    }
+
+    static @property void x(string v)
+    {
+        mX = v;
+    }
+
+    static alias x this;
+}
+
+// same as TStringRet, but for static @property functions
+struct TStringRetStatic
+{
+    static string mX;
+
+    static @property string x()
+    {
+        return mX;
+    }
+
+    static @property string x(string v)
+    {
+        mX = v;
+        return mX;
+    }
+
+    static alias x this;
+}
+
+// Test string (i.e. array) operators
+void testTString(T)()
+{
+    static if (typeid(T) is typeid(TString))
+    {
+        TString t;
+    }
+    else static if (typeid(T) is typeid(TStringStatic))
+    {
+        alias t = TStringStatic;
+    }
+    else
+    {
+        static assert(false, "Type is not supported");
+    }
+
+    t.x = "abc";
+    t.x ~= "def";
+    assert(t.mX == "abcdef");
+
+    // same as test above, but through the `alias this`
+    t = "abc";
+    t ~= "def";
+    assert(t.mX == "abcdef");
+}
+
+// This test is to verify that the setter @property function
+// returns a value if it is explicitly coded to do so
+void testTStringRet(T)()
+{
+    static if (typeid(T) is typeid(TStringRet))
+    {
+        TStringRet t;
+    }
+    else static if (typeid(T) is typeid(TStringRetStatic))
+    {
+        alias t = TStringRetStatic;
+    }
+    else
+    {
+        static assert(false, "Type is not supported");
+    }
+
+    string s;
+    t.x = "abc";
+    s = t.x ~= "def";
+    assert(s == "abcdef");
+
+    // same as test above, but through the `alias this`
+    t = "abc";
+    s = t ~= "def";
+    assert(s == "abcdef");
+}
+
+/**************************************************************
+* @ref property breakage test
+**************************************************************/
+struct TRefInt
+{
+    TRefInt[] mX;
+
+    this(TRefInt[] i)
+    {
+        mX = i;
+    }
+
+    @property ref TRefInt[] x()
+    {
+        return mX;
+    }
+
+    TRefInt opBinaryRight(string op)(TRefInt[] lhs) 
+    { 
+        mixin("return TRefInt(lhs "~op~" x);"); 
+    }
+}
+
+void testRefInt()
+{
+    // Ensure a ref @property is not rewritten
+    // This was an odd corner case found in dub's source code. We
+    // test specifically for it to avoid breakage
+    TRefInt tri1;
+    TRefInt tri2;
+    tri1.mX ~= tri2;
+    tri2.mX ~= tri1;
+    tri1.x ~= tri2; // tri2.opBinaryRight!("~=")(tri1.x)
+    assert(tri1.mX.length == 2);
+}
```

Comment:
Shouldn't this also not rewrite things for non-ref property? (with setter defined)
Better to make opBinaryRight do something different, so you can verify that it was called.